funopen() unavailable in gcc-mingw-w64-x86-64

In upgrading to libconfuse 3.1, I am now receiving the following error on mingw builds for Windows:
```
libconfuse.a(libconfuse_la-fmemopen.o): In function `fmemopen':
fmemopen.c:110: undefined reference to `funopen'
```
This can be reproduced on Debian/Ubuntu by installing the mingw crosscompiler:
```
sudo apt install gcc-mingw-w64-x86-64
cd confuse
CC=x86_64-w64-mingw32-gcc ./configure --enable-shared=no --disable-examples
make
```
Then try linking libconfuse.a to a trivial app. Since I never build the libconfuse examples in my Windows builds, I hadn't realized that they didn't compile with mingw, but in theory, they could show the issue too. If nothing else, you can see the warning the `funopen` isn't defined when compiling fmemopen.c.

I'm not completely sure how best to work around this issue. In the past when I've run into uses of fmemopen, it has been pretty easy to remove them, but I don't think that's the case here. However, I also don't know how one would implement fmemopen without funopen. Hence I'm posting here to see if there are any other ideas.

--- a/.gitignore
+++ b/.gitignore
@@ -1,5 +1,10 @@
 .gdb_history
 ABOUT-NLS
+GPATH
+GRTAGS
+GSYMS
+GTAGS
+ID
 aclocal.m4
 autom4te.cache
 config.h

--- a/configure.ac
+++ b/configure.ac
@@ -40,7 +40,7 @@ AC_CHECK_HEADERS([unistd.h string.h strings.h sys/stat.h windows.h])
 AC_C_CONST
 
 # Checks for library functions.
-AC_CHECK_FUNCS([fmemopen strcasecmp strdup strndup setenv unsetenv _putenv])
+AC_CHECK_FUNCS([fmemopen funopen strcasecmp strdup strndup setenv unsetenv _putenv])
 
 # Set conditional includes in Makefile.am
 AM_CONDITIONAL(MISSING_FMEMOPEN, [test "x$ac_cv_func_fmemopen" = "xno"])

--- a/configure.ac
+++ b/configure.ac
@@ -34,7 +34,7 @@ AM_GNU_GETTEXT_VERSION([0.18.2])
 
 # Checks for header files.
 AC_HEADER_STDC
-AC_CHECK_HEADERS([unistd.h string.h strings.h sys/stat.h])
+AC_CHECK_HEADERS([unistd.h string.h strings.h sys/stat.h windows.h])
 
 # Checks for typedefs, structures, and compiler characteristics.
 AC_C_CONST
@@ -44,6 +44,7 @@ AC_CHECK_FUNCS([fmemopen strcasecmp strdup strndup setenv unsetenv _putenv])
 
 # Set conditional includes in Makefile.am
 AM_CONDITIONAL(MISSING_FMEMOPEN, [test "x$ac_cv_func_fmemopen" = "xno"])
+AM_CONDITIONAL(WINDOWS_BUILD, [test "x$ac_cv_header_windows_h" = "xyes"])
 
 # Files to generate
 AC_CONFIG_FILES([Makefile \

--- a/examples/.gitignore
+++ b/examples/.gitignore
@@ -1,4 +1,5 @@
 *.o
+*.exe
 .deps
 .libs
 cfgtest

--- a/examples/Makefile.am
+++ b/examples/Makefile.am
@@ -1,5 +1,5 @@
 EXTRA_DIST      = simple.conf reread.conf ftp.conf test.conf nested.conf deprecated.conf
-noinst_PROGRAMS = simple ftpconf cfgtest cli nested deprecated addsec
+noinst_PROGRAMS = simple ftpconf cfgtest cli nested deprecated addsec parsebuf
 AM_CPPFLAGS     = -I$(top_srcdir)/src
 AM_LDFLAGS      = -L../src/
 LIBS            = $(LTLIBINTL)

--- a/examples/Makefile.am
+++ b/examples/Makefile.am
@@ -1,7 +1,13 @@
 EXTRA_DIST      = simple.conf reread.conf ftp.conf test.conf nested.conf deprecated.conf
-noinst_PROGRAMS = simple reread ftpconf cfgtest cli nested deprecated addsec
+noinst_PROGRAMS = simple ftpconf cfgtest cli nested deprecated addsec
 AM_CPPFLAGS     = -I$(top_srcdir)/src
 AM_LDFLAGS      = -L../src/
 LIBS            = $(LTLIBINTL)
 LDADD           = ../src/libconfuse.la
 CLEANFILES      = *~ \#*\#
+
+if !WINDOWS_BUILD
+noinst_PROGRAMS += reread
+else
+noinst_PROGRAMS += wincfgtest
+endif

--- a/examples/addsec.c
+++ b/examples/addsec.c
@@ -1,7 +1,15 @@
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
 #include <errno.h>
 #include <string.h>
-#include <unistd.h>
 #include <locale.h>
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+#ifdef HAVE_STDLIB_H
+#include <stdlib.h>
+#endif
 #include "confuse.h"
 
 cfg_t *cfg = NULL;

--- a/examples/addsec.c
+++ b/examples/addsec.c
@@ -45,8 +45,10 @@ int main(void)
 	cfg_t* sec;
 
 	/* Localize messages & types according to environment, since v2.9 */
+#ifdef LC_MESSAGES
 	setlocale(LC_MESSAGES, "");
 	setlocale(LC_CTYPE, "");
+#endif
 
 	read_config();
 	print_message();

--- a/examples/addsec.c
+++ b/examples/addsec.c
@@ -1,10 +1,10 @@
-#include <err.h>
+#include <errno.h>
 #include <string.h>
 #include <unistd.h>
 #include <locale.h>
 #include "confuse.h"
 
-cfg_t *cfg = 0;
+cfg_t *cfg = NULL;
 const char *config_filename = "./reread.conf";
 
 void read_config(void)
@@ -21,8 +21,10 @@ void read_config(void)
 	};
 
 	cfg = cfg_init(opts, CFGF_NONE);
-	if (cfg_parse(cfg, config_filename) != CFG_SUCCESS)
-		errx(1, "Failed parsing configuration!\n");
+	if (cfg_parse(cfg, config_filename) != CFG_SUCCESS) {
+		fprintf(stderr, "Failed parsing configuration: %s\n", strerror(errno));
+		exit(1);
+	}
 }
 
 void print_message()
@@ -59,7 +61,6 @@ int main(void)
 	print_message();
 
 	cfg_free(cfg);
-	cfg = 0;
 
 	return 0;
 

--- a/examples/ftpconf.c
+++ b/examples/ftpconf.c
@@ -110,8 +110,10 @@ int main(int argc, char **argv)
 	cfg_t *cfg;
 
 	/* Localize messages & types according to environment, since v2.9 */
+#ifdef LC_MESSAGES
 	setlocale(LC_MESSAGES, "");
 	setlocale(LC_CTYPE, "");
+#endif
 
 	cfg = parse_conf(argc > 1 ? argv[1] : "ftp.conf");
 	if (cfg) {

--- a/examples/parsebuf.c
+++ b/examples/parsebuf.c
@@ -1,0 +1,35 @@
+/* Very simple example for parse buf example */
+#include <string.h>
+#include <stdlib.h>
+#include "confuse.h"
+
+int main(void)
+{
+	cfg_t *cfg;
+	cfg_opt_t arg_opts[] = {
+		CFG_STR("value", "default", CFGF_NONE),
+		CFG_END()
+	};
+	cfg_opt_t opts[] = {
+		CFG_INT("delay", 3, CFGF_NONE),
+		CFG_STR("message", "This is a message", CFGF_NONE),
+		CFG_SEC("argument", arg_opts, CFGF_MULTI | CFGF_TITLE),
+		CFG_END()
+	};
+
+	char *buf =
+		" delay = 3\n"
+		"# message = 'asdfasfasfd tersf'\n"
+		" argument one { value = bar }\n"
+		" argument two { value=foo}\n";
+
+	cfg = cfg_init(opts, 0);
+	if (cfg_parse_buf(cfg, buf) != CFG_SUCCESS) {
+		fprintf(stderr, "Failed parsing configuration: %s\n", strerror(errno));
+		exit(1);
+	}
+	cfg_print(cfg, stdout);
+
+	return cfg_free(cfg);
+}
+

--- a/examples/parsebuf.c
+++ b/examples/parsebuf.c
@@ -1,4 +1,5 @@
 /* Very simple example for parse buf example */
+#include <errno.h>
 #include <string.h>
 #include <stdlib.h>
 #include "confuse.h"

--- a/examples/reread.c
+++ b/examples/reread.c
@@ -51,8 +51,10 @@ int main(void)
 	unsigned int i;
 
 	/* Localize messages & types according to environment, since v2.9 */
+#ifdef LC_MESSAGES
 	setlocale(LC_MESSAGES, "");
 	setlocale(LC_CTYPE, "");
+#endif
 
 	read_config();
 	signal(SIGHUP, sighandler);

--- a/examples/simple.c
+++ b/examples/simple.c
@@ -33,8 +33,10 @@ int main(void)
 	cfg_t *cfg;
 
 	/* Localize messages & types according to environment, since v2.9 */
+#ifdef LC_MESSAGES
 	setlocale(LC_MESSAGES, "");
 	setlocale(LC_CTYPE, "");
+#endif
 
 	/* set default value for the server option */
 	server = strdup("gazonk");

--- a/src/fmemopen.c
+++ b/src/fmemopen.c
@@ -14,7 +14,13 @@
  * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
  */
 
+#ifdef HAVE_CONFIG_H
+# include <config.h>
+#endif
+
 #include <stdio.h>
+
+#ifdef HAVE_FUNOPEN
 #include <stdlib.h>
 #include <memory.h>
 
@@ -109,6 +115,51 @@ FILE *fmemopen(void *buf, size_t len, const char *type)
 
 	return funopen(ops, readfn, writefn, seekfn, closefn);
 }
+#elif defined(HAVE_WINDOWS_H)
+#include <io.h>
+#include <fcntl.h>
+#include <windows.h>
+
+FILE *fmemopen(void *buf, size_t len, const char *type)
+{
+	int fd;
+	FILE *fp;
+	char tp[MAX_PATH - 13];
+	char fn[MAX_PATH + 1];
+	HANDLE h;
+
+	if (!GetTempPath(sizeof(tp), tp))
+		return NULL;
+
+	if (!GetTempFileName(tp, "confuse", 0, fn))
+		return NULL;
+
+	h = CreateFile(fn, GENERIC_READ | GENERIC_WRITE, 0, NULL,
+		       CREATE_ALWAYS, FILE_FLAG_DELETE_ON_CLOSE, NULL);
+	if (INVALID_HANDLE_VALUE == h)
+		return NULL;
+
+	fd = _open_osfhandle((intptr_t)h, _O_APPEND);
+	if (fd < 0) {
+		CloseHandle(h);
+		return NULL;
+	}
+
+	fp = fdopen(fd, "w+");
+	if (!fp) {
+		CloseHandle(h);
+		return NULL;
+	}
+
+	fwrite(buf, len, 1, fp);
+	rewind(fp);
+
+	return fp;
+}
+
+#else
+#error Sorry, this platform currently has no fmemopen() replacement.
+#endif
 
 /**
  * Local Variables:

